### ISSUE TITLE
DDC expands existing constants when they are embedded in other constants

### ISSUE THREAD

**Issue by nshahan** [dart-lang/sdk#37523]
Following on from issue https://github.com/dart-lang/sdk/issues/36535 there is still a problem of a code size explosion when a constant is embedded in another constant. Given the following dart code we will generate calls to the const list constructor every time `a` appears in `b`.

```dart
main() {
  const a = [1];
  const b = [a,a,a];
}

```

When we dump the dill file we can see that one constant are references the other. Is there a mechanism already built to help us recognize this in the backend and generate better code?

```
main = tes::main;
library from "file:///Users/nshahan/code/test_files/test_03.dart" as tes {

  static method main() → dynamic {
    const core::List<core::int> a = #C2;
    const core::List<core::List<core::int>> b = #C3;
  }
}
constants  {
  #C1 = 1
  #C2 = <core::int>[#C1]
  #C3 = <core::List<core::int>>[#C2, #C2, #C2]
}
```

**Comment by Markzipan** [comment]
After speaking with @fishythefish, we can construct that constants table as we walk the kernel AST. On visitXConstantReference, prepare an entry in our in-memory constant table, visiting it first if it hadn't yet been seen by us. On visitXConstant, just unroll it. 

The difficult part would be ensuring that imported and exported constants compare properly (with JS semantics).

**Comment by rakudrama** [comment]
The compilers are taking worst-case time exponential in the size of the program and the number of DAG nodes in the constant.

This can be observed by the failure to process the following 50-line program.
If I change main to use `n25`, the VM takes 7s, with `n27`, ~25s.
dart2js also has exponential compile time and heap which seems to be in walking the DAG (Constantifier in src/ir/visitors.dart).

So it looks like all our compilers need work in order to walk the constant DAGs efficiently.

Can we construct a general test where the kernel constant node is instrumented to flush out tools that walk a node too many times?

I suggest we open an issue for each compiler to handle several variants of the below file
1. Compile and run the code below without timeout
2. Compile and run deep DAGs of other constant expressions - Lists, Sets, Maps, interpolations etc.
3. Deep DAG that zig-zags between two libraries
4. Deep DAG that zig-zags between two libraries with deferred loading.

```dart
main() {
  print(n40.lengthA);
}
class Node {
  final Node a;
  final Node b;
  const Node(this.a, this.b);

  int get lengthA => a == null ? 0 : 1 + a.lengthA;
}

const n0 = Node(null, null);
const n1 = Node(n0, n0);
const n2 = Node(n1, n1);
const n3 = Node(n2, n2);
const n4 = Node(n3, n3);
const n5 = Node(n4, n4);
const n6 = Node(n5, n5);
const n7 = Node(n6, n6);
const n8 = Node(n7, n7);
const n9 = Node(n8, n8);
const n10 = Node(n9, n9);
const n11 = Node(n10, n10);
const n12 = Node(n11, n11);
const n13 = Node(n12, n12);
const n14 = Node(n13, n13);
const n15 = Node(n14, n14);
const n16 = Node(n15, n15);
const n17 = Node(n16, n16);
const n18 = Node(n17, n17);
const n19 = Node(n18, n18);
const n20 = Node(n19, n19);
const n21 = Node(n20, n20);
const n22 = Node(n21, n21);
const n23 = Node(n22, n22);
const n24 = Node(n23, n23);
const n25 = Node(n24, n24);
const n26 = Node(n25, n25);
const n27 = Node(n26, n26);
const n28 = Node(n27, n27);
const n29 = Node(n28, n28);
const n30 = Node(n29, n29);
const n31 = Node(n30, n30);
const n32 = Node(n31, n31);
const n33 = Node(n32, n32);
const n34 = Node(n33, n33);
const n35 = Node(n34, n34);
const n36 = Node(n35, n35);
const n37 = Node(n36, n36);
const n38 = Node(n37, n37);
const n39 = Node(n38, n38);
const n40 = Node(n39, n39);
```

**Comment by Markzipan** [comment]
@rakudrama I can't speak for Dart2JS, but the reason why DDK behaves poorly for that example is that we exponentially revisit every sub-constant. It would be pretty simple to store each Constant in some memoization table (using its deeply-comparable hashCode), which is what the CFE printer does. But it's not immediately clear to me whether this would work for the cross-module tests: 3) and 4). 

An exported Const1(Const2) should be identical to another Const2(cross-module C). Consider the following cross-module test:

`main.dart`:
```
import 'package:expect/expect.dart';

import 'def.dart';

main() {
  const v1 = A(B());
  const v2 = ab;
  const v3 = A(b);
  Expect.equals(v1, v2);
  Expect.equals(v2, v3);
}
```

`def.dart`:
```
class B {
  const B();
}

class A {
  final B b; 
  const A(this.b);
}

const ab = A(B());
const b = B();
```

`modules.yaml`:
```
dependencies:
  main: [def, expect]
flags:
  - constant-update-2018
```

This already gives me UnevaluatedConstants, so it memoizing stuff might just be "free" for now.

**Comment by nshahan** [comment]
@johnniwinther I'm still unsure about the expected CFE behavior here. If we ever encounter an unevaluated constant in a backend, should that be considered a bug in the CFE?

**Comment by vsmenon** [comment]
@johnniwinther - tentatively assigning this to you.  Stephen's example above seems to blow up all the backends.  We need to either handle this differently in the CFE or implement efficient dedup-ing in all the backends.

**Comment by aartbik** [comment]
Looked at this for DartVM: the final call to H.Canonicalize(instance) made from ConstantEvaluator::EvaluateConstant() is progressively getting slower and slower due to its recursive nature.

**Comment by vsmenon** [comment]
@Markzipan is working on DDC deduping.

**Comment by a-siva** [comment]
Fix in the VM for this case https://dart-review.googlesource.com/c/sdk/+/109318

**Comment by Markzipan** [comment]
Fix for DDK out at: https://dart-review.googlesource.com/c/sdk/+/109640

**Comment by Markzipan** [comment]
Update: there are a couple of staggering issues with the DDK implementation that will take a bit to hash out. Directly pulling out the consts is creating strange issues with DDK's built-in JS() function, and cyclic dependencies when building the SDK means we have to switch to a lazy loading scheme.

**Comment by Markzipan** [comment]
DDK update: We have a working change that lazily loads constants. However, it's caused a code size regression. Working on some optimizations that'll come in soon.

**Comment by Markzipan** [comment]
Turns out the code size "regression" was due to some sort of local glitch. Rebuilding the SDK actually gives us code size improvements. DDK changes are ready to land.

**Comment by Markzipan** [comment]
The DDK regression has been fixed - waiting to land.

**Comment by rakudrama** [comment]
The fix for the VM include new torture tests that need to be addressed in dart2js:
```
language_2/canonicalization_hashing_memoize_array_test
language_2/canonicalization_hashing_memoize_instance_test
language_2/canonicalization_hashing_shallow_collision_array_test
language_2/canonicalization_hashing_shallow_collision_instance_test
```
Two or three of these time out, depending on `--enable-experiment=constant-update-2018`.

There appear to be two reasons:
1. Under `--test-mode` dart2js forces the serialization used by multi-process compilation. This appears to serializes `ir.Constant`s without taking care of their DAG property.
2. `ir.Constant` has a poor hash code that causes Map collisions. The xor-ing in the ir.Constant hash codes should be replaced with something with better dispersal.

I can look at both of these on Tuesday.

**Comment by Markzipan** [comment]
We're encountering some bot failures that (mostly) aren't able to be reproduced locally. Consulting eng-prod before committing just in case the issue isn't bot-related.

**Comment by Markzipan** [comment]
The failing tests are due to: https://github.com/dart-lang/sdk/issues/37619

The 'dartdevc_test' dart build action uses an older version of the SDK to spit out various .dill files. Modifying pkg/dev_compiler/tool/build_pkgs.dart to pass the `--enable-experiment constant-update-2018` flag fixes the problem. 

So the test failures are due to weirdly configured DDC-specific tests. Looking into this.

**Comment by Markzipan** [comment]
These changes are ready to land but require the constant update to be enabled. Waiting on the re-merge before committing.
